zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
+        )
+        .optional()?
+    {
+        Some(i) if i < 0 => Err(SqliteClientError::CorruptedData(
+            "negative index".to_owned(),
+        )),
+        Some(i) => Ok(i),
+        None => Ok(-1),
+    }
+}
+
+/// Returns the last ephemeral address index in the given account that
+/// would not violate the gap invariant if used.
+pub(crate) fn last_safe_index(
+    conn: &rusqlite::Connection,
+    account_id: AccountId,
+) -> Result<u32, SqliteClientError> {
+    // The inner join with `transactions` excludes addresses for which
+    // `mined_in_tx` is NULL. The query also excludes addresses observed
+    // to have been mined in a transaction that we currently see as unmined.
+    // This is conservative in terms of avoiding violation of the gap
+    // invariant: it can only cause us to get to the end of the gap sooner.
+    let last_mined_index: i32 = match conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             JOIN transactions t ON t.id_tx = mined_in_tx
+             WHERE account_id = :account_id AND t.mined_height IS NOT NULL
+             ORDER BY address_index DESC
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
+        )
+        .optional()?
+    {
+        Some(i) if i < 0 => Err(SqliteClientError::CorruptedData(
+            "negative index".to_owned(),
+        )),
+        Some(i) => Ok(i),
+        None => Ok(-1),
+    }?;
+    Ok(u32::try_from(last_mined_index.saturating_add(GAP_LIMIT)).unwrap())
+}
+
+/// Utility function to return an `InclusiveRange<u32>` that starts at `i + 1`
+/// and is of length up to `n`. The range is truncated if necessary to end at
+/// the maximum valid address index, `i32::MAX`.
+///
+/// Precondition: `i >= -1 and n > 0`
+pub(crate) fn range_after(i: i32, n: i32) -> RangeInclusive<u32> {
+    assert!(i >= -1);
+    assert!(n > 0);
+    let first = u32::try_from(i64::from(i) + 1).unwrap();
+    let last = u32::try_from(i.saturating_add(n)).unwrap();
+    first..=last
+}
+
+/// Returns the ephemeral transparent IVK for a given account ID.
+pub(crate) fn get_ephemeral_ivk<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+) -> Result<EphemeralIvk, SqliteClientError> {
+    Ok(get_account(conn, params, account_id)?
+        .ok_or(SqliteClientError::AccountUnknown)?
+        .ufvk()
+        .and_then(|ufvk| ufvk.transparent())
+        .ok_or(SqliteClientError::UnknownZip32Derivation)?
+        .derive_ephemeral_ivk()?)
+}
+
+/// Returns a vector with all ephemeral transparent addresses potentially belonging to this wallet.
+/// If `for_detection` is true, this includes addresses for an additional GAP_LIMIT indices.
+pub(crate) fn get_reserved_ephemeral_addresses<P: consensus::Parameters>(
```

Comment:
Nonblocking: this API should probably provide an option for pagination by taking `Option<Page>` where 
```
struct Page {
    offset: u32,
    limit: u32
}
```

In general the way that we handle finding outputs for our addresses is awful, and we desperately need to revisit it.

Reply:
Out of scope for this PR.

Reply:
This is now implemented (using `Option<Range<u32>>`).